### PR TITLE
[Fix-10666] Workflow submit failed will still in memory and never retry

### LINKED ISSUE #10666: [Bug] [Master] Workflow submit failed will still in memory and never retry
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.

### What happened

When we execute a workflow, if the workflow submit failed in MasterServer, the WorkflowRunnable will always stay in memory, and it will not retry.
This is caused we didn't judge the submit status.

### What you expected to happen

When workflow submit failed, it will retry.

### How to reproduce

Submit the workflow, and restart the mysql.

### Anything else

_No response_

### Version

dev

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by ruanwenjun**
## Purpose of the pull request

close #10666 

## Brief change log
- Add a daemon thread to retry the submit failed processInstance

### CODE REVIEW

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/exception/MasterException.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package org.apache.dolphinscheduler.server.master.exception;
```
Comment (SbloodyS) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/exception/MasterException.java:18]: Missing license header.

Comment (ruanwenjun) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/exception/MasterException.java:18]: Done

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/MasterSchedulerService.java [hunk 2]
```diff
@@ -164,38 +172,53 @@ private void scheduleWorkflow() throws InterruptedException {
...
+                workflowExecuteRunnable::call, workflowExecuteThreadPool);
+            workflowSubmitFuture.thenAccept(workflowSubmitStatue -> {
+                if (WorkflowSubmitStatue.FAILED == workflowSubmitStatue) {
+                    // submit failed
+                    processInstanceExecCacheManager.removeByProcessInstanceId(processInstance.getId());
+                    stateWheelExecuteThread.removeProcess4TimeoutCheck(processInstance.getId());
+                    submitFailedProcessInstances.add(processInstance);
+                }
+            });
+            logger.info("Master schedule service started workflow instance");
```
Comment (caishunfeng) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/MasterSchedulerService.java:214]:
```suggestion
            logger.info("Master schedule service started workflow instance, id:{}, name:{} ", processInstance.getId(), processInstance.getName());
```

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/MasterSchedulerService.java [hunk 3]
```diff
@@ -164,38 +172,53 @@ private void scheduleWorkflow() throws InterruptedException {
...
-            } finally {
-                LoggerUtils.removeWorkflowInstanceIdMDC();
+            submitProcessInstance(processInstance);
+        }
+    }
+
+    private void submitProcessInstance(@NonNull ProcessInstance processInstance) {
+        try {
+            LoggerUtils.setWorkflowInstanceIdMDC(processInstance.getId());
+            logger.info("Master schedule service starting workflow instance");
```
Comment (caishunfeng) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/MasterSchedulerService.java:189]: add processInstance id or name

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/MasterSchedulerService.java [hunk 4]
```diff
@@ -164,38 +172,53 @@ private void scheduleWorkflow() throws InterruptedException {
...
+                    stateWheelExecuteThread.removeProcess4TimeoutCheck(processInstance.getId());
+                    submitFailedProcessInstances.add(processInstance);
+                }
+            });
+            logger.info("Master schedule service started workflow instance");
+
+        } catch (Exception ex) {
+            processInstanceExecCacheManager.removeByProcessInstanceId(processInstance.getId());
+            stateWheelExecuteThread.removeProcess4TimeoutCheck(processInstance.getId());
+            logger.info("Master submit workflow to thread pool failed, will remove workflow runnable from cache manager", ex);
```
Comment (caishunfeng) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/MasterSchedulerService.java:219]: add processInstance id or name

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/MasterSchedulerService.java (hunk 2 above)
Comment (ruanwenjun) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/MasterSchedulerService.java:214]: We already use [MDC](https://logback.qos.ch/manual/mdc.html) to set the processInstanceId by `LoggerUtils.setWorkflowInstanceIdMDC(processInstance.getId());` so the log will bring this id, we don't need to set this in every log.

**Review by caishunfeng — DISMISSED** [review]
LGTM

**Comment by caishunfeng** [comment]
Hi @ruanwenjun please resolve the conflicts.

**Comment by ruanwenjun** [comment]
@lenboo

**Review by SbloodyS — APPROVED** [review]
+1

**Pull request merged**